[10.0][ADD] Sendgrid modules

ecino opened the pull request:
These new modules allow to use Sendgrid as mail provider and benefit from tracking functionalities as well as templating functionalities from Sendgrid.

ecino committed:
Add Sendgrid module from v9

ecino committed:
Add Sendgrid Mass Mailing from v9

ecino committed:
[10.0] Migrate sendgrid modules

CO-1192 The mail are now correctly sent to all recipients

pedrobaeza commented on mail_sendgrid/README.rst:
```diff
@@ -0,0 +1,120 @@
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/205/9.0
```
s/9.0/10.0

pedrobaeza commented on mail_sendgrid/__init__.py:
```diff
@@ -0,0 +1,14 @@
+# -*- encoding: utf-8 -*-
```
Please use current template and put correctly `coding`: https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L1-L3

ecino replied:
Fixed all headers

ecino commented:
Don't know why pylint fails on importing lxml... it is installed in .travis.yml file https://github.com/OCA/social/blob/10.0/.travis.yml#L13

pedrobaeza commented:
It's a general problem: https://github.com/OCA/maintainer-quality-tools/issues/467

ecino committed:
Sendgrid code corrections for v10 + Add tests

ecino commented:
Tests are fixed, the module is stable, we are using it in production in version 10 since some time now. We only miss a review to merge.

ecino commented:
Is someone willing to review?

pedrobaeza commented:
Are you able to improve the coverage for having everything green? If so, I will optimistically merge the PR although no reviews.

ecino commented:
I already tried to improve the coverage but the targeted value is very high: here is the code putting down the coverage, but I don't know if we could include those parts in the tests and if it would be relevant, what do you think?

https://codecov.io/gh/OCA/social/src/d97bd6f63d54ee8fa029ea8c08ed2482b40571d6/mail_sendgrid/controllers/sendgrid_event_webhook.py

https://codecov.io/gh/OCA/social/src/d97bd6f63d54ee8fa029ea8c08ed2482b40571d6/mail_sendgrid/controllers/json_request.py

hbrunn commented:
@pedrobaeza why would we do such a thing? Of course no merging without review, and also no merging with red CI, even if it's "only" tests.

hbrunn commented:
@ecino I'd propose you review some PRs and ask the author of those to return the favor

hbrunn commented:
...for testing your controllers, you can use [HttpCase](https://github.com/OCA/OCB/blob/11.0/odoo/tests/common.py#L247)

pedrobaeza commented:
@hbrunn I reviewed his work on previous version and I'll review also before merging this one, but no one is able to review such a big thing, and it's a good work from not a newcomer (@ecino has made already several contributions). It's for not discouraging him from more contributions.

hbrunn commented:
I understand the reasoning, still I think if we lower our standard everyone loses. Anyways, now realizing most of the code is already reviewed and merged, I could review the rest easily and we've one +1. @ecino what I often do in such cases it to tell the review which commits are the interesting ones, that allows reviewers to trust the rest and focus on new stuff

ecino commented:
@hbrunn Thanks for the tips, it's always useful to improve myself. In that case the relevant commits are the 3 last (after migration of code for version 10)

simahawk commented on mail_sendgrid/__manifest__.py:
```diff
@@ -0,0 +1,25 @@
+    'category': 'Social Network',
+    'author': 'Compassion CH, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'website': 'http://www.compassion.ch',
```
https://github.com/OCA/social

simahawk commented on mail_sendgrid/controllers/sendgrid_event_webhook.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class SendgridTrackingController(MailTrackingController):
+    """
```
one line docstring

simahawk commented on mail_sendgrid/models/email_lang_template.py:
```diff
@@ -0,0 +1,22 @@
+    _name = 'sendgrid.email.lang.template'
+
+    email_template_id = fields.Many2one('mail.template', 'E-mail Template')
+    lang = fields.Selection('_lang_get', 'Language', required=True)
```
`selection='_select_lang'`

simahawk commented on mail_sendgrid/models/email_template.py:
```diff
@@ -0,0 +1,76 @@
+class EmailTemplate(models.Model):
+    _inherit = 'mail.template'
+
+    ##########################################################################
```
is this needed?

simahawk commented on mail_sendgrid/models/email_template.py:
```diff
@@ -0,0 +1,76 @@
+    @api.multi
+    def update_substitutions(self):
+        self.ensure_one()
+        new_substitutions = list()
```
`= []`

simahawk commented on mail_sendgrid/models/email_template.py:
```diff
@@ -0,0 +1,76 @@
+            res_ids = [res_ids]
+        substitutions = self.substitution_ids.filtered(
+            lambda s: s.lang == self.env.context.get('lang', 'en_US'))
+        substitution_vals = {res_id: list() for res_id in res_ids}
```
```python
substitution_vals = {}.fromkeys(res_ids, [])
```
or even better
```python
from collection import defaultdict
substitution_vals = defaultdict(list)
```

ecino replied:
+1 for defaultdict, I didn't new this library!

simahawk commented on mail_sendgrid/models/email_tracking.py:
```diff
@@ -0,0 +1,169 @@
+    """
+    _inherit = 'mail.tracking.email'
+
+    click_count = fields.Integer(compute='_compute_clicks', store=True)
```
readonly?

simahawk commented on mail_sendgrid/models/email_tracking.py:
```diff
@@ -0,0 +1,169 @@
+    @api.depends('tracking_event_ids')
+    def _compute_clicks(self):
+        for mail in self:
+            mail.click_count = len(mail.tracking_event_ids.filtered(
```
do you expect many records? if yes, go for a `search` instead of `filtered`

simahawk commented on mail_sendgrid/models/email_tracking.py:
```diff
@@ -0,0 +1,169 @@
+        # OK, event type is valid
+        return True
+
+    def _db_verify(self, event):
```
better to prefix this w/ `_sendgrid` as you did - nicely - w/ the others (unless this comes from the mail_tracking module ;P)

simahawk commented on mail_sendgrid/models/email_tracking.py:
```diff
@@ -0,0 +1,169 @@
+        ts = event.get('timestamp', False)
+        try:
+            ts = float(ts)
+        except:
```
```except ValueError```

simahawk commented on mail_sendgrid/models/email_tracking.py:
```diff
@@ -0,0 +1,169 @@
+
+    def _sendgrid_metadata(self, sendgrid_event_type, event, metadata):
+        # Get sendgrid timestamp when found
+        ts = event.get('timestamp', False)
```
default is `None` in any case

simahawk commented on mail_sendgrid/models/email_tracking.py:
```diff
@@ -0,0 +1,169 @@
+                    tracking = self._sendgrid_tracking_get(event)
+                    if not tracking:
+                        res = 'ERROR: Tracking not found'
+                if res == 'OK':
```
do you need this further "if"?

simahawk commented on mail_sendgrid/models/mail_mail.py:
```diff
@@ -0,0 +1,252 @@
+    """
+    _inherit = 'mail.message'
+
+    ##########################################################################
```
same as above

simahawk commented on mail_sendgrid/models/mail_mail.py:
```diff
@@ -0,0 +1,252 @@
+
+    def _compute_events(self):
+        for email in self:
+            email.tracking_event_ids = email.tracking_email_ids.mapped(
```
same story: if you expect many records use a search

ecino replied:
search instead of mapped? for better performance?

simahawk commented on mail_sendgrid/models/mail_mail.py:
```diff
@@ -0,0 +1,252 @@
+            email.send_method = send_method
+
+
+class OdooMail(models.Model):
```
MailMail

simahawk commented on mail_sendgrid/models/mail_mail.py:
```diff
@@ -0,0 +1,252 @@
+                 'tracking_email_ids.state')
+    def _compute_tracking(self):
+        for email in self:
+            email.click_count = sum(email.tracking_email_ids.mapped(
```
use `write` to write once on both fields

ecino replied:
I thought write method should not be used in compute methods?

hbrunn replied:
it absolutely shouldn't and wouldn't work here anyways. What the colleague probably means is [.update()](https://github.com/OCA/OCB/blob/10.0/odoo/models.py#L5010)

simahawk commented on mail_sendgrid/models/mail_mail.py:
```diff
@@ -0,0 +1,252 @@
+    ##########################################################################
+    #                             PRIVATE METHODS                            #
+    ##########################################################################
+    def _prepare_sendgrid_data(self):
```
if possible break this into smallest methods

simahawk commented on mail_sendgrid/models/mail_tracking_event.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+# © 2017 Emanuel Cino - <ecino@compassion.com>
```
use `Copyright` instead of symbol

simahawk commented on mail_sendgrid/models/sendgrid_template.py:
```diff
@@ -0,0 +1,96 @@
+                self.detected_keywords = ';'.join(keywords)
+
+    @api.model
+    def update(self):
```
:stop_sign: `update` is an[ internal method](https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L5010) and is used to update in memory record w/ writing/creating things.

simahawk commented on mail_sendgrid/tests/test_mail_sendgrid.py:
```diff
@@ -0,0 +1,233 @@
+
+
+class TestMailSendgrid(HttpCase):
+    def setUp(self):
```
if possible use `setUpClass` to create records that are always used: boosts test perf a lot ;)

ecino replied:
Not working with HttpCase? I get errors that cls.env is not defined

simahawk commented on mail_sendgrid/views/email_template_view.xml:
```diff
@@ -0,0 +1,31 @@
+        <field name="model">mail.template</field>
+        <field name="inherit_id" ref="mail.email_template_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='email_from']/../.." position="after">
```
`/ancestor::page` should be better

simahawk commented on mail_sendgrid/views/sendgrid_email_view.xml:
```diff
@@ -0,0 +1,77 @@
+            <field name="body_html" position="attributes">
+                <attribute name="widget">html</attribute>
+            </field>
+            <xpath expr="//field[@name='attachment_ids']/.." position="after">
```
same here

simahawk commented on mail_sendgrid/wizards/email_template_preview.py:
```diff
@@ -0,0 +1,23 @@
+
+    @api.onchange('res_id')
+    @api.multi
+    def on_change_res_id(self):
```
`onchange_[...]`

ecino replied:
No, because the parent method in odoo/addons/mail has this name

simahawk replied:
sorry, my bad, I missed the `super` call ;P

simahawk commented on mail_sendgrid_mass_mailing/README.rst:
```diff
@@ -0,0 +1,61 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/social/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/social/issues/new?body=module:%20mail_sendgrid_mass_mailing%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
update this w/ [latest README](https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst)

simahawk commented on mail_sendgrid_mass_mailing/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    'category': 'Social Network',
+    'author': 'Compassion CH, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'website': 'http://www.compassion.ch',
```
https://github.com/OCA/social

simahawk commented on mail_sendgrid_mass_mailing/models/mass_mailing.py:
```diff
@@ -0,0 +1,133 @@
+            mailing = self.with_context(lang=lang)
+            author = mailing.mass_mailing_campaign_id.user_id.partner_id \
+                or self.env.user.partner_id
+            composer_values = {
```
better to split this to `_send_mail_get_composer_values` or alike

simahawk commented on mail_sendgrid_mass_mailing/tests/test_mass_mailing.py:
```diff
@@ -0,0 +1,169 @@
+
+
+class TestMailSendgrid(TransactionCase):
+    def setUp(self):
```
same: try using `setUpClass`

ecino replied:
I tried with it but I get many errors in tearDown (DB no such savepoint)

simahawk replied:
you have to use `common.SavepointCase` as base class instead.

ecino replied:
Yes I tried that, but then I don't know what the errors are about in the tearDown method.

simahawk replied:
which errors? do you have a traceback?

ecino replied:
Ok, that was my bad, I forgot to add the `@classmethod` decorators and change tearDown to tearDownClass as well.

I still couldn't change the test of Sendgrid because of the `HttpCase` class.

simahawk reviewed (changes requested):
wow, a long one :sweat_smile:

ecino commented:
@simahawk Thank you very much for your helpful review! I corrected most of what you found.
@pedrobaeza I updated the tests with more coverage. Unfortunately the coverage is not yet enough... very hard to match 98.8%!

rafaelbn commented:
@simahawk please could you unblock this PR? :smile:

rafaelbn commented:
If there is not any inconvenient I will merge this today evening.

simahawk reviewed (approved):
@ecino @rafaelbn sorry, went off my radar :stuck_out_tongue_winking_eye:

ecino commented:
@rafaelbn Can we merge?

pedrobaeza commented:
Can you squash a bit the commit history before merging? I see several commits with same text and other not very relevant.

ecino commented:
@pedrobaeza Done
